### ISSUE TITLE
question: exporting word specific tags

### ISSUE THREAD

**Issue by st32lthx** [jgm/pandoc#2542]
Is it possible to export documents with cues for word styles ? For example, let's say in my reference `docx` file there is a style called `cool style`. Is there a way to add some cues in the markdown file so that when it is exported it gets the correct styles ? Something like:

`file.md`
```markdown

This is a sentence with {special style}(cool-style)
```

Thanks !

**Comment by jgm** [comment]
+++ Amin Meyghani [Nov 16 15 19:46 ]:
>   Is it possible to export documents with cues for word styles ?

Currently, no.  But in principle it should be possible to
set things up so that `<span class="foo">bar</span>` gets
rendered with style "Span-foo" (or something like that).
Alternatively there could be a special attribute,
`<span data-word-style="foo">`.

**Comment by aminmeyghani** [comment]
Awesome ! Any pointers on how to get that work ? Would it be something like:

`file.md` -> `css, html` -> `docx`

Thanks!

**Comment by jgm** [comment]
I can implement it (it would be direct md -> docx).  It's just a matter of
figuring out what scheme makes most sense (mapping all
classes on spans to correspondingly named Word styles,
if they are defined, might make most sense).

+++ Amin Meyghani [Nov 16 15 20:13 ]:

**Comment by aminmeyghani** [comment]
That would be really cool! Btw, I am still memorized by the robustness of the tool, it's just so robust, it's unreal ....

**Comment by mb21** [comment]
A proposed by [massifrg on pandoc-discuss](https://groups.google.com/forum/#!topic/pandoc-discuss/BVusehQGrQE), this issue could be extended to:

1. cover not only DOCX, but also ODT and ICML
2. and cover not only Inlines (spans) but also Blocks (divs) (i.e. not only "character styles" but also "paragraph styles")

btw, do the styles really need to be defined in the reference word file or ICML template to trigger the rendering of those spans/divs as styles? Seems like a lot of work parsing those... Maybe a prefix like `style-` is enough, e.g. `<span class="style-foo">bar</span>`, or a custom attribute: `<span data-style="foo">bar</span>`.

**Comment by mb21** [comment]
In LaTeX those divs/spans could be rendered as custom environments/commands. This was actually proposed by @hadley in https://github.com/jgm/pandoc/issues/168, or do you think the concepts are not analogous enough?

**Comment by jgm** [comment]
+++ Mauro Bieg [Dec 31 15 09:38 ]:
>   In LaTeX those divs/spans could be rendered as custom
>   environments/commands. This was actually proposed by [1]@hadley in
>   [2]#168, or do you think the concepts are not analogous enough?

I want to avoid generating environments and commands that
aren't defined (and similarly for styles in Word and ICML).
If we parse the styles, and thus know what is available,
that may not be a big problem.  In LaTeX it's harder,
because commands and environments may be defined in
included packages.  The idea of having a special prefix
like `style-` might be a good one.

**Comment by bpj** [comment]
I have written several fairly simple filters, both bespoke and more general, which insert raw latex before and/or after spans or divs based on classes or custom attributes. ([An example](https://gist.github.com/bpj/f67d274c9a0688014ad9#file-pandoc-expand-simple-py))  Since that, as @jgm said, requires that the inserted commands/environments are or get defined somewhere it is probably better to leave that to filters so that the user is in full control, since they must include the necessary macro definitions at the same time.  Too much 'automagic' with such things will be troublesome.

**Comment by massifrg** [comment]
I'm trying to pass the "style-..." classes to ICML with a filter that injects RawInlines into ICML.
Example:
```html
  <span class="style-foo">custom style</span>
```
becomes (pandoc --filter myfilter.pl -f markdown -t icml)
```xml
  <CharacterStyleRange AppliedCharacterStyle="CharacterStyle/foo">
    <Content>custom style</Content>
  </CharacterStyleRange>
```
I'm not following jgm's rule of not defining styles that are not present in the target document, because it's a custom filter for personal and exploratory purposes (*for now*).

Anyway that was enough to stumble upon a problem: **nested character styles**.
What would you get trying to convert this?
```html
  <span class="style-foo">custom <span class="style-bar">style</span></span>
```
In HTML you can nest a &lt;span&gt; inside another &lt;span&gt;, multiple times.

In ICML you can't nest a &lt;CharacterStyleRange&gt; inside another &lt;CharacterStyleRange&gt;.

That is true for DOCX too, where you can't nest a &lt;w:r&gt; inside another &lt;w:r&gt;.

In ODT you can nest a nest a &lt;text:span&gt; inside another &lt;text:span&gt;, but you won't be able to do it with AOO Writer or LO Writer. You can do it by editing the XML directly, and you obtain what you would expect: the styles merge when they are applied together. Perhaps when you reopen it in LO Writer one of the &lt;text:span&gt; would be considered *the* character style and the others would be transformed into local modifications.
The question is well summarized in this [forum](http://forum.openoffice.org/en/forum/viewtopic.php?f=7&t=64859).

Before passing custom styles to target documents, there's a problem of "flattening" the character styles' depth to one level, and there's not a unique (and perhaps neither a good) way to do it.
It's a problem you face even with a simple filter, before you think to add this feature to pandoc.

Anyway I'm trying to write that filter, because my documents are in HTML and passing styles to InDesign is a great feature. I'll let you know. That could be a contribution useful for a more general solution.

**Comment by mb21** [comment]
I guess this issue was resolved through https://github.com/jgm/pandoc/pull/3070

**Comment by jgm** [comment]
I think that's right, I'll close this.

**Comment by monobjorn** [comment]
@massifrg Would you like to share your filter? I would really appreciate it.

**Comment by rufuspollock** [comment]
To clarify:

* Passing styles to ICML from Markdown #2663 was merged into here (=> this issue is about both word and markdown ...)
* This was then closed by by #3070
* However I'm not clear that #3070 actually supports passing styles to icml from markdown
* If it does not should we reopen #2663

**Comment by mb21** [comment]
I think the issue tracking custom styles for more writer is #2106

**Comment by rufuspollock** [comment]
@mb21 so #2663 should really be open or closed in favour of #2106 (not this issue)?
